core/src/main/scala/spinal/core/sim/package.scala

```diff
@@ -356,80 +358,65 @@ package object sim {
     }
   }
 
-  /**
-    * Add implicit function to Bits
-    */
-  implicit class SimBitsPimper(bt: Bits) {
+  protected abstract class RandomizableBitVector(bt: BitVector, longLimit: Int) {
+    protected val width = bt.getWidth
 
-    def randomize(): Unit = {
-      val width = bt.getWidth
-      if(width < 64){
-        bt #= Random.nextLong() & ((1l << width) - 1)
-      }else {
-        bt #= BigInt(width, Random)
+    def randomize(): BigInt = {
+      if (width < longLimit) {
+        val l = randomizedLong()
+        bt #= l
+        l
+      } else {
+        val bi = randomizedBigInt()
+        bt #= bi
+        bi
       }
     }
 
     def randomizedBigInt() = {
-      val width = bt.getWidth
       BigInt(width, Random)
     }
+
     def randomizedLong() = {
-      val width = bt.getWidth
       assert(width < 64)
       Random.nextLong() & ((1l << width) - 1)
     }
+
     def randomizedInt() = {
-      val width = bt.getWidth
       assert(width < 32)
       Random.nextInt() & ((1 << width) - 1)
     }
   }
 
   /**
-    * Add implicit function to UInt
-    */
-  implicit class SimUIntPimper(bt: UInt) {
-
-    def randomize(): Unit = {
-      val width = bt.getWidth
-      if(width < 64){
-        bt #= Random.nextLong() & ((1l << width) - 1)
-      }else {
-        bt #= BigInt(width, Random)
-      }
-    }
+   * Add implicit function to Bits
+   */
+  implicit class SimBitsPimper(bt: Bits) extends RandomizableBitVector(bt, 64)
 
-    def randomizedBigInt() = {
-      val width = bt.getWidth
-      BigInt(width, Random)
-    }
-    def randomizedLong() = {
-      val width = bt.getWidth
-      assert(width < 64)
-      Random.nextLong() & ((1l << width) - 1)
-    }
-    def randomizedInt() = {
-      val width = bt.getWidth
-      assert(width < 32)
-      Random.nextInt() & ((1 << width) - 1)
-    }
-  }
+  /**
+   * Add implicit function to UInt
+   */
+  implicit class SimUIntPimper(bt: UInt) extends RandomizableBitVector(bt, 64)
 
 
   /**
-    * Add implicit function to SInt
-    */
-  implicit class SimSIntPimper(bt: SInt) {
+   * Add implicit function to SInt
+   */
+  implicit class SimSIntPimper(bt: SInt) extends RandomizableBitVector(bt, 65) {
+    override def randomizedLong(): Long = {
+      assert(width <= 64)
+      val shift = 64 - width
+      (Random.nextLong << shift) >> shift
```

Comment:
I'm not sute I understand what you mean. We need an assert here since it's still user facing code, and the width to assert here differs from the unsigned case?

Reply:
`super.randomized…` already performs the same assertion doesn't it?

Reply:
If `longLimit` is inclusive it is possible to use it but anyway there is the same issue with `randomizedInt`.